### DISCUSSION TITLE
Async handlers don't work if they begin with `let`/`const` or an `if` statement

### DISCUSSION THREAD

**Discussion by willcosgrove** [alpinejs/alpine#3298] [7. Bugs (archived)]
### Discussed in https://github.com/alpinejs/alpine/discussions/3118

<div type='discussions-op-text'>

<sup>Originally posted by **willcosgrove** August 25, 2022</sup>
It looks like the issue is caused by wrapping the expression for safe right hand side evaluation:

https://github.com/alpinejs/alpine/blob/4df4e4f12aa929e69cde0cd0fde65ac624b8e5d8/packages/alpinejs/src/evaluator.js#L68-L77

On line 76, if the above conditions return true, the expression gets wrapped in `(() => ...)()` but that wrapping function is not async, and so any `await` calls will cause a syntax error.</div>

**Comment by willcosgrove** [comment]
Here's an example of the problem: https://codepen.io/willcosgrove/pen/rNdgjRv

**Comment by markjaquith** [comment]
I wonder if line 76 could just be:

```
? `await (async () => { ${expression} })()`
```

**Comment by willcosgrove** [comment]
I think that that would fix it. I don't know if there's any downside to always wrapping with an async function.

**Comment by fweth** [comment]
As I wrote in the linked discussion, please also fix the patterns.

`/^\s*(let|const)[\s[{]/` captures more legal JS strings which logically start with `let`/`const`, e.g. `let[x,y]=[2,3]` or `const{x,y}={x:2,y:3}` or a string starting with a newline, which I do sometimes for longer statements, like this:
```html
<button
  @click="
    let str = 'hello';
    console.log(str);
  "
><button>
```
The current pattern for `if` seems fine, but a little bit expensive as it looks until finding a closing paranthesis (which doesn't even need to match the opening one).  Also, I think `\n` is already included in `\s`.  I'd replace it with `/^\s*if[\s(]/`

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/alpinejs/alpine/discussions/1423
